### PR TITLE
Support configuring timeout for outbound requests

### LINKED ISSUE #76: Support Configurable Timeouts
Admin SDK uses Google Java HTTP client to make outbound HTTP connections. But we currently do not set the connection and read timeout for outbound requests. These need to be configurable via `FirebaseOptions` or a similar API.

### PR DESCRIPTION
**Pull request by hiranya911**
Resolves #76

### CODE REVIEW

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 1]
```diff
@@ -132,6 +138,26 @@ public JsonFactory getJsonFactory() {
     return jsonFactory;
   }
 
+  /**
+   * Returns the connection timeout in milliseconds, which is applied to outgoing REST calls
+   * made by the SDK.
+   *
+   * @return Connection timeout in milliseconds. 0 indicates an infinite timeout.
+   */
+  public int getConnectTimeout() {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/FirebaseOptions.java]: We discussed this during the API review. We agree that eventually we want to support per-component and perhaps even per-operation level timeouts. But that's a major change, and will probably require rethinking the overall design of the SDK. The current change gives developers a working solution until we get there.

I have updated the API docs to indicate that the timeouts do not apply to RTDB and Firestore.

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 2]
```diff
@@ -321,6 +351,28 @@ public Builder setThreadManager(ThreadManager threadManager) {
       return this;
     }
 
+    /**
+     * Sets the connection timeout for outgoing HTTP (REST) calls made by the SDK.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/FirebaseOptions.java]: ConnectTimeout is the correct and more commonly used terminology. For example: https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#setConnectTimeout(int)

It does apply to the initial TCP connection  handshake. I've updated the text to be consistent with the method name.

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 3]
```diff
@@ -321,6 +351,28 @@ public Builder setThreadManager(ThreadManager threadManager) {
...
+     *
+     * @param connectTimeout Connection timeout in milliseconds. Must not be negative.
+     * @return This <code>Builder</code> instance is returned so subsequent calls can be chained.
+     */
+    public Builder setConnectTimeout(int connectTimeout) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/FirebaseOptions.java]: Yes.

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 4]
```diff
@@ -36,6 +36,7 @@
 import com.google.firebase.auth.UserRecord.UpdateRequest;
 import com.google.firebase.auth.internal.FirebaseTokenFactory;
 import com.google.firebase.auth.internal.FirebaseTokenVerifier;
+import com.google.firebase.internal.FirebaseRequestInitializer;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/FirebaseAuth.java]: Good catch :) Removed.

File: src/main/java/com/google/firebase/internal/FirebaseRequestInitializer.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2017 Google Inc.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/internal/FirebaseRequestInitializer.java]: Done

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 6]
```diff
@@ -132,6 +138,26 @@ public JsonFactory getJsonFactory() {
     return jsonFactory;
   }
 
+  /**
+   * Returns the connection timeout in milliseconds, which is applied to outgoing REST calls
```
Comment (schmidt-sebastian) [inline src/main/java/com/google/firebase/FirebaseOptions.java]: s/connection/connect/ here as well.

Comment (hiranya911) [reply src/main/java/com/google/firebase/FirebaseOptions.java]: Done

**Pull request merged**